### ISSUE TITLE
ENH: `einsum_path` should be backend-agnostic 

### ISSUE THREAD

**Issue by leofang** [numpy/numpy#21379]
### Proposed new feature or change:

I am creating this feature request wearing two hats (CuPy and cuQuantum Python).

Currently, `numpy.einsum_path()` supports the same calling convention as `numpy.einsum()` with regard to the input tensors and the einsum expression. So either of these is accepted:
- `numpy.einsum_path(arr1, sublist1, arr2, sublist2, ...)`
- `numpy.einsum_path(einsum_str, *(arr1, arr2, ...))`

But either of this is quite an overkill, as `einsum_path()` does not care about the array contents or most of the array attributes other than their shapes (strides aren't even taken into account during the path computation). Therefore, unlike `numpy.einsum()` it can be made *backend-agnostic*: As long as the backend array (NumPy, CuPy, PyTorch, ...) has the concept of shapes, they can all delegate to `numpy.einsum_path()` to compute the path without reinventing the wheel. 

For this purpose, it'd be nice if `numpy.einsum_path()` can additionally support the following calling conventions:
- `numpy.einsum_path(shape1, sublist1, shape2, sublist2, ...)`
- `numpy.einsum_path(einsum_str, *(shape2, shape2, ...))`

(picking one of them is also fine, if it's easier.)

On the **CuPy** side, this naive delegation to implement `cupy.einsum_path()`
```python
def einsum_path(subscripts, *operands, optimize='greedy'):
    return numpy.einsum_path(subscripts, *operands, optimize=optimize)
```
does not work because it'd be trapped in an infinite `__array_function__` dispatching loop. As a result, currently CuPy has to copy a ton of the `numpy.einsum_path` internal code. This is not fun and has maintenance overhead (we can't co-evolve with the latest improvement in NumPy).

On the **cuQuantum Python** side, we'd like to provide a path we calculate to `numpy.einsum_path()` for pretty-printing the contraction information so that we don't have to reinvent this bit of machinery (or being outdated from NumPy),
```python
path = cuquantum.contract_path(...)
path, info = numpy.einsum_path(..., optimize=['einsum_path', *path])
```
However, as we support multiple array backends this is a problem (ex: we have to create temporary NumPy arrays if the input tensors are from PyTorch or CuPy in order to do this).

**Comment by charris** [comment]
@dgasmith Ping.

**Comment by leofang** [comment]
@dgasmith's `opt_einsum.contract_path` supports this capability via the `shapes` option, so I am pretty sure it's doable 😄 https://dgasmith.github.io/opt_einsum/api_reference/#opt_einsumcontract_path

**Comment by leofang** [comment]
I am wondering, would it be possible to disable the `__array_function__` dispatch? This would immediately support all array containers that implement the `.shape` attribute, without introducing any new calling convention or options.

**Comment by shoyer** [comment]
I agree that the `einsum_path` API is should probably have been desigend to work on shapes instead of arrays, but changing this is a tricky from a backwards compatibility perspective. Tuples are currently interpreted as 1d arrays.

One way to work around this that I've used in the past is to create "dummy" arrays of the appropriate shapes with array broadcasting. Something like `numpy.einsum_path(numpy.broadcast_to(0, shape1), sublist1, numpy.broadcast_to(0, shape2), sublist2, ...)` should do the trick. Perhaps this is worth adding to the docs for `einsum_path`?  The use of array broadcasting means that NumPy never needs to actually create potentially large arrays in memory.

**Comment by dgasmith** [comment]
For more advanced functionality such as those in cuQuantum it is recommended to use `opt_einsum` as the diversity of algorithms continues to grow for advanced use cases. As @shoyer points out, I would be hesitant to change too much of the `einsum_path` interface and the available solution will work.

**Comment by leofang** [comment]
If there's any NumPy 2.0 planning tracker, I'd like to have this considered. cc: @seberg

**Comment by seberg** [comment]
I am planning on making a place for such a list soon, let me tag as "major release" for now.   That said, I would hope we can find a solution without a need for 2.0.

I suspect we could get away with passing through non-NumPy array objects (with a `.shape` attribute and maybe also `__array_priority__` or so).
If the idea is to accept tuples as shapes, then I think a 2.0 may be a good excuse to at least cut short on a `FutureWarning` (which we could add at any point).
